src/main/java/org/folio/fqm/service/EntityTypeService.java

```diff
@@ -70,7 +77,54 @@ public Optional<EntityType> getEntityTypeDefinition(UUID entityTypeId) {
    */
   @Transactional(readOnly = true)
   public ColumnValues getColumnValues(UUID entityTypeId, String columnName, @Nullable String searchText) {
-    String fql = String.format(COLUMN_VALUE_SEARCH_FQL, columnName, searchText == null ? "" : searchText);
+    searchText = searchText == null ? "" : searchText;
+
+    EntityType entityType = entityTypeRepository.getEntityTypeDefinition(entityTypeId)
+      .orElseThrow(() -> new EntityTypeNotFoundException(entityTypeId));
+    EntityTypeColumn column = entityType.getColumns()
+      .stream()
+      .filter(col -> Objects.equals(columnName, col.getName()))
+      .findFirst()
+      .orElseThrow(() -> new ColumnNotFoundException(entityType.getName(), columnName));
+
+    if (column.getValues() != null) {
+     return getColumnValuesFromEntityTypeColumn(column, searchText);
+    }
+
+    if (column.getValueSourceApi() != null) {
+      return getColumnValuesFromApi(column, searchText);
+    }
+
+    return getColumnValuesFromEntityType(entityTypeId, columnName, searchText);
+
+  }
+
+  private ColumnValues getColumnValuesFromEntityTypeColumn(EntityTypeColumn column, String searchText) {
+    var filteredColumns = column.getValues().stream()
+      .filter(valueWithLabel -> valueWithLabel.getLabel().contains(searchText))
+      .toList();
+    return new ColumnValues().content(filteredColumns);
+  }
+
+  private ColumnValues getColumnValuesFromApi(EntityTypeColumn column, String searchText) {
+    String rawJson = columnValueClient.get(column.getValueSourceApi().getPath());
+    DocumentContext parsedJson = JsonPath.parse(rawJson);
+    List<String> values = parsedJson.read(column.getValueSourceApi().getValueJsonPath());
+    List<String> labels = parsedJson.read(column.getValueSourceApi().getLabelJsonPath());
+
+    List<ValueWithLabel> results = new ArrayList<>(values.size());
+    for (int i = 0; i < values.size(); i++) {
+      String value = values.get(i);
+      String label = labels.get(i);
+      if (label.contains(searchText)) {
+        results.add(new ValueWithLabel().value(value).label(label));
+      }
+    }
+    return new ColumnValues().content(results);
+  }
```

Comment:
So in functional programming languages, there tends to be a function called `zip` that combines 2 collections together into a collection of pairs. Oracle actually had a `Stream.zip()` in the pre-releases of Java 8, but decided to remove it before releasing. If we still had that, then this would have been soooooo much cleaner as a stream... This would have looked something like
```java
return values.stream()
  .zip(labels.stream())
  .filter(valueAndLabel -> valueAndLabel.getSecond().contains(searchText))
  .map(valueAndLabel -> new ValueWithLabel().value(valueAndLabel.getFirst()).label(valueAndLabel.getSecond())
  .toList()
```